src/expr.rs

```diff
@@ -0,0 +1,97 @@
+use std::fmt;
+
+#[derive(Debug, Clone)]
+pub enum Expr {
+    Unary(UnaryOp, Box<Expr>),
+    Literal(Literal),
+    Binary(Box<Expr>, BinaryOp, Box<Expr>),
+    Grouping(Box<Expr>),
+}
+
+impl fmt::Display for Expr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let message = match self {
+            Expr::Unary(operator, expression) => format!("({} {})", operator, expression),
+            Expr::Literal(lit) => format!("{}", lit),
+            Expr::Binary(left, operator, right) => format!("({} {} {})", operator, left, right),
+            Expr::Grouping(expression) => format!("(group {})", expression),
+        };
+        write!(f, "{}", message)
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum BinaryOp {
+    Plus,
+    Minus,
+    Star,
+    Slash,
+    EqualEqual,
+    BangEqual,
+    Less,
+    LessEqual,
+    Greater,
+    GreaterEqual,
+}
+
+impl fmt::Display for BinaryOp {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let message = match self {
+            BinaryOp::Plus => "+",
+            BinaryOp::Minus => "-",
+            BinaryOp::Star => "*",
+            BinaryOp::Slash => "/",
+            BinaryOp::EqualEqual => "==",
+            BinaryOp::BangEqual => "!=",
+            BinaryOp::Less => "<",
+            BinaryOp::LessEqual => "<=",
+            BinaryOp::Greater => ">",
+            BinaryOp::GreaterEqual => ">=",
+        };
+
+        write!(f, "{}", message)
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum UnaryOp {
+    Minus,
+    Bang,
+}
+
+impl fmt::Display for UnaryOp {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        return match self {
+            UnaryOp::Minus => write!(f, "-"),
+            UnaryOp::Bang => write!(f, "!"),
+        };
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Literal {
+    Str(String),
```

Comment:
Again, here you'd ideally not copy `String`s over. Optimizing `Literal` is probably important because it's going to be used 1000s of times in a program, and doing heap allocations 1000s of times isn't great for performance.

Depending on how deep you want to go with this parser, here are some strategies to think about memory:

### General advice for `String`

So the "general rust advice" would be that this could be `&'a str` but then `Literal` will need a lifetime, which means `Expr` will and the whole thing bubbles up into a bit of a mess. `&str` is also 16 bytes, which means `Literal` will be at-least 16 bytes.

But potentially we could store less...

### Parser specific advice

Another way to think this is to think about the specific modelling of a parser itself. `Literal` does not sit in a vacuum. During the execution of your parser, the parser will be given the source to interpret, and the Parser will hold that. Think about what `&str` is and what data it might contain; if you were to write it as a struct:

```rust
struct Str {
  pointer_to_offset_of_real_string_in_memory: usize,
  length_of_this_slice_of_the_real_string: usize
}
```

There are two things that make this interesting, and why a specialisation might be worth pursuing in this one case. Firstly, `pointer_to_offset_of_real_string_in_memory` will point to a memory location where the underlying `String` is, at an offset of the desired string. Forgetting the offset part for a second, we don't necessarily _need_ to know where the `String` is in memory because the `Parser` has that, and `Literal` only really gets used in the context of the parser. So `Literal::Str` could be:

```rust
enum Literal {
  Str({ offset: usize, length: usize })
}
```

But we haven't gained anything here. This is still a 16byte enum. However let's think about `usize`. A `usize` is 8 bytes on 64bit systems, because it needs to be able to address up to ~18 Exabytes of memory (it's 4 bytes on 32-bit systems which means it can only address ~4gb of memory). I don't know about you, but I don't think I've ever encountered source code in the Gigabyte range, let alone Terra/Peta/Exabytes. So what's a realistic minimum? A `u8` is 1 byte but the max value is 255, and I've definitely seen a file bigger than 255 characters. A `u16`'s max value is `65535`, or 65MB... theoretically there are source files bigger than 65mb, especially if you're trying to parse big bundled source, but it would be very rare. And then we have the u32 which, again, is 4gb. 

So given these numbers, maybe the offset doesn't need to be `usize`, maybe it can be a `u32` and an appropriate limitation of your parser is that it won't parse files bigger than 4gb. Changing them both to `u32` slashes the struct from 16 to 8bytes:

```rust
enum Literal {
  Str({ offset: u32, length: u32 })
}
```

We _could_ apply the same technique to the `length` too; how likely is it that a program is going to _only_ have 1 string and that string is going to be 4gb? However, let's shrink `length` down to u16 and stick it in a file and run with Rust's `-Zprint-type-sizes` flag, to get size information:

```sh
$ cat file.rs
  #![allow(unused)]
  fn main() {
      enum Literal {
          Str(u32, u16),
      }
  }

$ rustc +nightly -Zprint-type-sizes file.rs

  ...
  print-type-size type: `main::Literal`: 8 bytes, alignment: 4 bytes
  print-type-size     variant `Str`: 6 bytes
  print-type-size         field `.0`: 4 bytes
  print-type-size         field `.1`: 2 bytes
  print-type-size     end padding: 2 bytes
  ...
```

Huh... so `field .0` (our u32) is 4 bytes. `field .1` is 2 bytes, and our `Str` variant is indeed 6 bytes. But the struct as a whole is 8 bytes, not 6... Rust added 2 bytes of padding!?

Rust will pad structs to specific memory boundaries, for efficiency in memory layout. It's just added 2 bytes of end padding to struct, making the whole thing 8 bytes. It tells us why: `alignment: 4 bytes`. It's trying to align the struct into 4 byte memory bounds. This means our optimization can't go lower than 4byte fields because Rust will just pad to multiples of 4.

So we've got two options:

1. We make `Str(u32, u16)` and live with the fact that Rust will pad it to 8 bytes anyway. This might save us later if we need to cram a couple bytes more data into this field, for example if we wanted to encode the quote style the string uses we could have `Str(u32, u16, u8)` and Rust would only pad the extra byte, meaning `Str(u32, u16, u8)` would still be 8 bytes.
2. We make `Str(u32, u32)` and move on with our lives.

Reply:
> So the "general rust advice" would be that this could be `&'a str` but then `Literal` will need a lifetime, which means `Expr` will and the whole thing bubbles up into a bit of a mess.

This is what I was running into and _couldn't_ figure out what would be best. It's a good shout because I was gonna come back to this but forgot.

The comment is great (as always :pray:) I'll take a further look over the weekend!

Reply:
Actually, one question: If I did all this, how do I take the offset and length and translate that to a string? Do I have to pass in a reference of the source code or something?

Reply:
Yeah you'd have to have a method, perhaps on parser:

```rust
fn get_source_at_position(&self, start: u32, end: u32) -> &'a str {
  self.source[start..end]
}
```

Reply:
What if the parser just has `Vec<Token>` and not actually the source code? 👀 

Should the parser have access to the actual source code?

Reply:
Why would the parser end up with tokens but no source?

Reply:
That's what the book walks you through. First you run the tokenizer on the source code and then you pass those tokens to the parser to get expressions.

https://craftinginterpreters.com/parsing-expressions.html#the-parser-class

Reply:
Ah, okay. In that case the Tokenizer could house the source code instead.

Reply:
Right, and I think it does right now? As a `&'a str`. Is it just a matter or "ordering" things? Instead of running the tokenizer and then the parser with it's output I should pass the whole tokenizer to the parser? Or have some sort of wrapping struct called compiler? Something that would allow the parser access to the source code (probably via the tokenizer).

Reply:
For hdx the parser and tokenizer both get `&'a str` (remember that’s a cheap pointer so fine to pass around). But the tokenizer only lives as long as reading the source and Tokens themselves have no lifetimes bound, then the parser sticks around to build the AST which also has the same lifetime

Reply:
Makes sense! I think for now I'm gonna add a TODO and start working on the interpreter and statements as I started working on that and got excited.

So for now I'm gonna merge this and create a new PR for the interpreter.

Reply:
```suggestion
    // TODO: Convert this to a `&'a str`. See https://github.com/koddsson/codecrafters-interpreter-rust/pull/1#discussion_r1684154612
    Str(String),
```